Pipeline/PublishSharedInterfaces/PublishPublicInterfaces.groovy

```diff
@@ -0,0 +1,204 @@
+
+if(!jsonOutputFile.exists()){
+	println("** Build report data at $properties.workDir/BuildReport.json not found")
+	System.exit()
+}
+
+def buildReport= BuildReport.parse(new FileInputStream(jsonOutputFile))
+
+// parse build report to find the build result meta info
+def buildResult = buildReport.getRecords().findAll{it.getType()==DefaultRecordFactory.TYPE_BUILD_RESULT}[0];
+def dependencies = buildReport.getRecords().findAll{it.getType()==DefaultRecordFactory.TYPE_DEPENDENCY_SET};
+
+// parse build report to find the build outputs to be deployed.
+println("** Find and display Public Interfaces in the build report.")
+
+// find all the public interfaces which got copied to the pds
+def publicCopies= buildReport.getRecords().findAll{
+	it.getType()==DefaultRecordFactory.TYPE_COPY_TO_PDS &&
+			!it.getOutputs().findAll{ o ->
+				o.deployType == deployTypeName
+			}.isEmpty()
+}
+
+// only run logic, if we found Public Interfaces
+if (publicCopies.size()!=0){
+
+	setLock(lockFile)
+
+	String targetDir = properties.targetHfsDirectory +"/"+properties.application
+	def gitHash
+	publicCopies.each {
+		Path originalPath = Paths.get("${it.source}");
+		Path targetPath =  Paths.get(targetDir)
+		Path targetFile = targetPath.resolve(originalPath.getFileName())
+
+		Files.createDirectories(targetPath)
+		println("* Copying   " + originalPath.getFileName() + " to $targetDir")
+		Files.copy(originalPath, targetFile, StandardCopyOption.REPLACE_EXISTING);
+		// set file tag
+		currentTag = processCmd("chtag -p " + originalPath).split("\\s+")[1];
+		processCmd("chtag -t -c $currentTag " + targetFile.toString())
+		gitHash = getCurrentGitHash(originalPath.getParent().toString())
+	}
+
+	//git opertions - git status + add
+	processCmd("git -C $targetDir status")
+	processCmd("git -C $targetDir add .")
+	processCmd("git -C $targetDir status")
+	
+	//git commit
+	ArrayList nextCmd=new ArrayList<String>()
+	nextCmd.add("git")
+	nextCmd.add("-C")
+	nextCmd.add(targetDir)
+	nextCmd.add("commit")
+	nextCmd.add("-m")
+	nextCmd.add("\"${properties.application}:${gitHash}\"")
+	processCmd(nextCmd)
+	
+	//git push
+	processCmd("git -C $targetDir push")
+
+    // Placeholder: IEBCOPY Files to shared library
+
+    //release lock
+    releaseLock(lockFile)
+}
+else {
+	println("*! No Public Copybooks found")
+}
+
+def setLock(Path sharedDir){
+	println("** Setting lock file to avoid collisions." )
+	def lockAttempts = 0
+
+	while(lockAttempts <  10){
+		try{
+			Files.createFile(sharedDir.resolve(lockingFile))
+			return
+		} catch (FileAlreadyExistsException e){
+			println("*! $lockingFile exits. Waiting 5 seconds. Attempt : $lockAttempts")
+			lockAttempts++
+			sleep(5000) // sleep for 5 seconds
+		}
+	}
+
+	println("*! Could not obtain lock for publishing files. Exiting." )
+	System.exit(0)
+}
+
+def releaseLock(Path sharedDir){
+	println("* Releasing lock file." )
+	Files.deleteIfExists(sharedDir.resolve(lockingFile))
+}
```

Comment:
How to handle the case where my build failed between  setLock/releaseLock ?
The lock file will always be there. I will have to delete it manually.

Reply:
You are right, in case of a nightly build there will be nobody to release the lock.